// @keyauthjs/client/src/utils/embedBuilder.ts
interface EmbedOptionsBase {
    color?: number;
    description?: string;
    timestamp?: string;
    title?: string;
    url?: string;
}
export interface EmbedOptions extends EmbedOptionsBase {
    author?: EmbedAuthorOptions;
    fields?: Array<EmbedField>;
    footer?: EmbedFooterOptions;
    image?: EmbedImageOptions;
    thumbnail?: EmbedImageOptions;
}
interface EmbedAuthorOptions extends EmbedAuthorBase {
    iconURL?: string;
}
interface EmbedAuthorBase {
    name: string;
    url?: string;
}
interface EmbedField {
    inline?: boolean;
    name: string;
    value: string;
}
interface EmbedFooterOptions extends EmbedFooterBase {
    iconURL?: string;
}
interface EmbedFooterBase {
    text: string;
}
interface EmbedImageOptions {
    url: string;
}
export default class EmbedBuilder {
    private _content: string | undefined;
    private _embeds: EmbedOptions[] | undefined;
    constructor({
        content,
        embeds,
    }: {
        content?: string | undefined;
        embeds?: EmbedOptions[] | undefined;
    }) {
        this._content = content;
        this._embeds = embeds;
    }
    toStrting() {
        const message = {
            content: this._content,
            embeds: this._embeds,
        };
        return JSON.stringify(message);
    }
    toJSON() {
        const message = {
            content: this._content,
            embeds: this._embeds,
        };
        return message;
    }
}
