// @keyauthjs/client/src/utils/helpers.ts
// Import necessary modules and types
import type { DownloadResponse, Subscription, Info } from "../types";
import { writeFile } from "fs";

/**
 * Calculates the time elapsed since a given date or timestamp.
 *
 * @param {Date | string} timeSince - The date or timestamp to calculate the time since.
 * @returns {string | 0} - A string describing the time elapsed or 0 if it's less than a minute.
 */
export function timeSince(timeSince: Date | string): string | 0 {
    // Get the current date and time
    const currentDate = new Date();

    // Convert the input to a Date object or multiply it by 1000 if it's in seconds
    const joinDate = new Date(
        timeSince instanceof Date ? timeSince : (timeSince as any) * 1000,
    );

    // Calculate the time difference in milliseconds
    const timeDifference = (currentDate as any) - (joinDate as any);

    // Calculate time units
    const seconds = Math.floor(timeDifference / 1000);
    const minutes = Math.floor(seconds / 60);
    const hours = Math.floor(minutes / 60);
    const days = Math.floor(hours / 24);

    // Return the appropriate string based on the elapsed time
    if (days > 0) {
        return `${days} day${days !== 1 ? "s" : ""} ago`;
    } else if (hours > 0) {
        return `${hours} hour${hours !== 1 ? "s" : ""} ago`;
    } else if (minutes > 0) {
        return `${minutes} minute${minutes !== 1 ? "s" : ""} ago`;
    } else {
        // If it's less than a minute, return 0
        return 0;
    }
}

/**
 * Converts a Unix timestamp to a local Date object.
 *
 * @param {string} timestampString - The Unix timestamp as a string.
 * @returns {Date} - A Date object representing the local date and time.
 */
export function convertUnixTimestampToLocalDate(timestampString: string): Date {
    // Parse the input timestamp string as an integer
    const unixTimestamp = parseInt(timestampString);

    // Convert the Unix timestamp to milliseconds and create a Date object
    return new Date(unixTimestamp * 1000); // Convert seconds to milliseconds
}

/**
 * Calculates the time remaining until a given expiry date or timestamp.
 *
 * @param {Date | string} expiry - The expiry date or timestamp.
 * @returns {string} - A string describing the time remaining.
 */
export function timeUntilExpiry(expiry: Date | string): string {
    // Get the current date and time
    const currentDate = new Date();

    // Convert the expiry input into a Date object (if it's not already)
    const expiryDate = new Date(
        expiry instanceof Date ? expiry : (expiry as any) * 1000,
    ); // Convert seconds to milliseconds

    // Calculate the time difference between the expiry and current date
    const timeDifference = (expiryDate as any) - (currentDate as any);

    // Calculate seconds, minutes, hours, and days from the time difference
    const seconds = Math.floor(timeDifference / 1000);
    const minutes = Math.floor(seconds / 60);
    const hours = Math.floor(minutes / 60);
    const days = Math.floor(hours / 24);

    // Construct a human-readable string describing the time remaining
    if (days > 0) {
        return `${days} day${days !== 1 ? "s" : ""} left`;
    } else if (hours > 0) {
        return `${hours} hour${hours !== 1 ? "s" : ""} left`;
    } else if (minutes > 0) {
        return `${minutes} minute${minutes !== 1 ? "s" : ""} left`;
    } else {
        return `${seconds} second${seconds !== 1 ? "s" : ""} left`;
    }
}

/**
 * Converts a downloaded response to a string.
 *
 * @param {DownloadResponse} downloadResponse - The response containing downloadable content.
 * @returns {string} - The downloaded content as a string.
 */
export function downloadToString(downloadResponse: DownloadResponse): string {
    // Convert the hexadecimal content of the download response into a byte array
    const byteArray = new Uint8Array(
        downloadResponse.contents
            .match(/.{1,2}/g)! // Split content into pairs of hexadecimal digits
            .map((byte) => parseInt(byte, 16)), // Convert each pair to a byte value
    );

    // Decode the byte array into a string
    const decodedString = new TextDecoder().decode(byteArray);

    // Return the downloaded content as a string
    return decodedString;
}

/**
 * Downloads content to a file.
 *
 * @param {DownloadResponse} downloadResponse - The response containing downloadable content.
 * @param {string} name - The name of the file.
 * @param {string} fileType - The type or extension of the file.
 * @param {string} location - The location where the file will be saved.
 */
export function downloadToFile(
    downloadResponse: DownloadResponse,
    name: string,
    fileType: string,
    location: string,
) {
    // Convert the hexadecimal content of the download response into a byte array
    const byteArray = new Uint8Array(
        downloadResponse.contents
            .match(/.{1,2}/g)! // Split content into pairs of hexadecimal digits
            .map((byte) => parseInt(byte, 16)), // Convert each pair to a byte value
    );

    // Write the byte array to a file
    writeFile(
        `${location}/${name}.${fileType}`.replace("//", "/"), // Build the file path
        byteArray, // Write the byte array
        "utf8", // Encoding format
        (err) => {
            if (err) {
                // Handle errors if any
                console.error("Error writing to file:", err);
            } else {
                // File write successful
                console.log(
                    `Data has been written to ${`${location}/${name}.${fileType}`.replace(
                        "//",
                        "/",
                    )}`,
                );
            }
        },
    );
}

/**
 * Converts timestamps in user data to local Date objects.
 *
 * @param {Info} data - User information object containing timestamps.
 * @returns {Info} - User information object with converted timestamps.
 */
export function convertTimestampsToLocalDates(data: Info): Info {
    // Create a copy of the input data to avoid modifying the original object
    const convertedUser = { ...data };

    // Create an array to store converted subscription objects
    const convertedSubs: Subscription[] = [];

    // Convert the 'createdate' timestamp to a local Date object
    convertedUser.createdate = convertUnixTimestampToLocalDate(
        data.createdate as string,
    );

    // Convert the 'lastlogin' timestamp to a local Date object
    convertedUser.lastlogin = convertUnixTimestampToLocalDate(
        data.lastlogin as string,
    );

    // Iterate through the subscriptions and convert their 'expiry' timestamps
    if (data.subscriptions)
        data.subscriptions.forEach((sub) => {
            const convertedSub = { ...sub };

            // Convert the 'expiry' timestamp to a local Date object
            convertedSub.expiry = convertUnixTimestampToLocalDate(
                convertedSub.expiry as string,
            );

            // Add the converted subscription to the array
            convertedSubs.push(convertedSub);
        });

    // Replace the original subscriptions with the converted array
    convertedUser.subscriptions = convertedSubs;

    // Return the user information object with converted timestamps
    return convertedUser;
}
