// @keyauthjs/client/src/utils/logger.ts
// Import necessary modules and types
import winston, { Logger, transports } from "winston";

// Logger configuration
export interface keyauthLogger {
    active?: boolean;
    level?: "error" | "warning" | "info" | "debug" | "dev";
    name?: string;
}

/**
 * CustomLogger class for creating and configuring a custom logger based on Winston.
 * Used in the logging for keyauth ClientAPI
 */
export default class CustomLogger {
    private _logger: Logger;
    private _name: string | undefined;

    /**
     * Constructs a CustomLogger instance.
     * @param {keyauthLogger} options - The logger options.
     * @param {string} options.name - The logger name.
     * @param {boolean} options.active - Whether the logger is active or not.
     * @param {string} options.level - The logging level (e.g., "error", "info").
     */
    constructor(options: keyauthLogger) {
        this._name = options.name;

        // Defining custom log levels and colors
        const customLevels = {
            error: 0,
            warning: 1,
            info: 2,
            debug: 3,
        };

        // Define custom log colors
        winston.addColors({
            error: "red",
            warning: "yellow",
            info: "blue",
            debug: "gray",
        });

        // Create the Winston logger instance with custom levels and desired configurations
        this._logger = winston.createLogger({
            // Control whether the logger is silent or not
            silent: options.active === true ? false : true ?? true,
            levels: customLevels,
            // Set the logging level (default to "error")
            level: options.level ?? "error",
            format: winston.format.combine(
                // Apply color to console output
                winston.format.colorize(),
                winston.format.timestamp(),
                // Define the log message format
                winston.format.printf(
                    ({ timestamp, level, message, name, tag }) => {
                        return `${timestamp} [${name}] [${tag}|${level}]: ${message}`;
                    },
                ),
            ),
            transports: [
                new transports.Console(), // Log to console
            ],
        });
    }

    /**
     * Log an error message.
     * @param {string} tag - The log tag.
     * @param {string} message - The error message.
     */
    error(tag: string, message: string) {
        this._logger.log("error", message, { name: this._name, tag });
    }

    /**
     * Log a warning message.
     * @param {string} tag - The log tag.
     * @param {string} message - The warning message.
     */
    warning(tag: string, message: string) {
        this._logger.log("warning", message, { name: this._name, tag });
    }

    /**
     * Log an info message.
     * @param {string} tag - The log tag.
     * @param {string} message - The info message.
     */
    info(tag: string, message: string) {
        this._logger.log("info", message, { name: this._name, tag });
    }

    /**
     * Log a debug message.
     * @param {string} tag - The log tag.
     * @param {string} message - The debug message.
     */
    debug(tag: string, message: string) {
        this._logger.log("debug", message, { name: this._name, tag });
    }
}
