// @keyauthjs/client/src/types/index.ts
// Interfaces for various data structures

import EventEmitter from "events";
import { keyauthLogger } from "../utils/logger";
import TypedEmitter from "typed-emitter";

/**
 * The application information can be found: https://keyauth.cc/app/
 */
export interface App {
    /**
     * The application name
     */
    name: string;
    /**
     * The application version [defaults to 1.0]
     */
    ver: string;
    /**
     * The applations owners owner ID
     */
    ownerid: string;
    // secret:string
}

/**
 * Information about a session
 */
export interface Session {
    /**
     * The current session ID
     */
    id: string;
    /**
     * If the current session is a new session or not
     */
    new: boolean;
    /**
     * If the current session is validated or not
     */
    validated: boolean;
}

/**
 * Base response structure
 */
export interface BaseResponse {
    /**
     * If the response was successful or not
     */
    success: boolean;
    /**
     * The message from the response
     */
    message: string;
    time: number;
    nonce?: string;
}

// Response after initializing
export interface InitResponse extends BaseResponse {
    sessionid?: string;
    newSession?: boolean;
}

export type METADATA = Record<string, any>;

// Response after a login attempt
export interface LoginResponse extends BaseResponse {
    metaData?: METADATA;
    info?: Info;
}

// Response after logging out
export interface LogoutResponse extends BaseResponse {}

// Configuration options
export interface ClientOptions {
    ratelimit?: { maxTokens: number; refillRate: number };
    convertTimes?: boolean;
    baseUrl?: string;
    logger?: Omit<keyauthLogger, "tag" | "name">;
}

// Parameters for initializing a session
export interface InitParams extends App {
    type: EVENT_TYPE.INIT;
}

// Parameters for a login request
export interface LoginParams {
    type: EVENT_TYPE.LOG_IN;
    username: string;
    pass: string;
    sessionid: string;
    hwid?: string;
}

// Parameters for a logout request
export interface LogoutParams {
    type: EVENT_TYPE.LOG_OUT;
    sessionid: string;
}

// Information about a subscription
export interface Subscription {
    subscription: string;
    key: string | null;
    expiry: string | Date;
    timeleft: number;
    level: string;
}

// User information
export interface Info {
    username: string;
    subscriptions: Subscription[];
    ip: string;
    hwid: string | null;
    createdate: string | Date;
    lastlogin: string | Date;
}

// User information with optional metadata
export interface User extends Info {
    metaData?: METADATA;
}

// Response after a registration attempt
export interface RegisterResponse extends BaseResponse {
    info?: Info;
    metaData?: METADATA;
}

// Parameters for user registration
export interface RegisterParams {
    type: EVENT_TYPE.REGISTER;
    username: string;
    pass: string;
    key: string;
    sessionid: string;
    email?: string;
}

// Response after license verification
export interface LicenseResponse extends BaseResponse {
    info?: Info;
}

// Parameters for license verification
export interface LicenseParams {
    type: EVENT_TYPE.LICENSE;
    key: string;
    sessionid: string;
    email?: string;
}

// Response after changing a username
export interface ChangeUsernameResponse extends BaseResponse {}

// Parameters for changing a username
export interface ChangeUsernameParams {
    type: EVENT_TYPE.CHANGE_USERNAME;
    newUsername: string;
    sessionid: string;
}

// Response after banning a user
export interface BanResponse extends BaseResponse {}

// Application information
export interface AppInfo {
    numUsers: number;
    numOnlineUsers: number;
    numKeys: number;
    version: string;
    customerPanelLink: string;
}

// Parameters for fetching statistics
export interface FetchStatsParams {
    type: EVENT_TYPE.FETCH_STATS;
    sessionid: string;
}

// Response after fetching statistics
export interface FetchStatsResponse extends BaseResponse {
    appinfo: AppInfo;
}

// Parameters for banning a user
export interface BanParams {
    type: EVENT_TYPE.BAN;
    sessionid: string;
    reason: string;
}

// Response after checking blacklist
export interface CheckBlacklistResponse extends BaseResponse {}

// Parameters for checking blacklist
export interface CheckBlacklistParams {
    type: EVENT_TYPE.CHECK_BLACKLIST;
    sessionid: string;
    hwid: string;
}

// Response after a session check
export interface CheckResponse extends BaseResponse {}

// Parameters for a session check
export interface CheckParams {
    type: EVENT_TYPE.CHECK;
    sessionid: string;
}

// Response after fetching online users
export interface FetchOnlineUsersResponse extends BaseResponse {
    users: {
        credential: string;
    }[];
    count: number;
}

// Parameters for fetching online users
export interface FetchOnlineUsersParams {
    type: EVENT_TYPE.FETCH_ONLINE;
    sessionid: string;
}

// Response after a password reset request
export interface ForgotPasswordResponse extends BaseResponse {}

// Parameters for a password reset request
export interface ForgotPasswordParams {
    type: EVENT_TYPE.FORGOT_PASSWORD;
    email: string;
    sessionid: string;
    username: string;
}

// Response after a log request
export interface LogResponse extends BaseResponse {}

// Parameters for a log request
export interface LogParams {
    type: EVENT_TYPE.LOG;
    sessionid: string;
    pcuser: string;
    message: string;
}

// Response after an upgrade request
export interface UpgradeResponse extends BaseResponse {}

// Parameters for an upgrade request
export interface UpgradeParams {
    type: EVENT_TYPE.UPGRADE;
    sessionid: string;
    username: string;
    key: string;
}

// Response after a webhook request
export interface WebhookResponse extends BaseResponse {
    data: any;
}

// Parameters for a webhook request
export interface WebhookParams {
    type: EVENT_TYPE.WEBHOOK;
    sessionid: string;
    webid: string;
    body?: string;
    params?: string;
    conttype?: "application/json" | string | undefined;
}

// Response after a file download request
export interface DownloadResponse extends BaseResponse {
    contents: string;
}

// Parameters for a file download request
export interface DownloadParams {
    type: EVENT_TYPE.DOWNLOAD;
    sessionid: string;
    fileid: string;
}
export class KeyauthEventEmitter extends (EventEmitter as new () => TypedEmitter<EventMap>) {
    constructor() {
        super();
    }
}
export type EventType =
    | "init"
    | "login"
    | "logout"
    | "register"
    | "license"
    | "fetchStats"
    | "ban"
    | "changeUsername"
    | "checkblacklist"
    | "check"
    | "file"
    | "fetchOnline"
    | "forgot"
    | "chatget"
    | "getvar"
    | "log"
    | "chatsend"
    | "setvar"
    | "upgrade"
    | "webhook"
    | "var"
    // Added custom types
    | "request"
    | "response"
    | "metadata"
    | "error"
    | "session"
    | "instance"
    | "ratelimit";

// Enum defining various event types
export enum EVENT_TYPE {
    INIT = "init",
    LOG_IN = "login",
    LOG_OUT = "logout",
    REGISTER = "register",
    LICENSE = "license",
    FETCH_STATS = "fetchStats",
    BAN = "ban",
    CHANGE_USERNAME = "changeUsername",
    CHECK_BLACKLIST = "checkblacklist",
    CHECK = "check",
    DOWNLOAD = "file",
    FETCH_ONLINE = "fetchOnline",
    FORGOT_PASSWORD = "forgot",
    CHAT_GET = "chatget",
    GET_VAR = "getvar",
    LOG = "log",
    CHAT_SEND = "chatsend",
    SET_VAR = "setvar",
    UPGRADE = "upgrade",
    WEBHOOK = "webhook",
    VAR = "var",
    // Added custom types
    REQUEST = "request",
    RESPONSE = "response",
    METADATA = "metadata",
    ERROR = "error",
    SESSION = "session",
    INSTANCE = "instance",
    RATE_LIMIT = "ratelimit",
}
export type EventMap = {
    init: (data: InitResponse) => void;

    login: (data: LoginResponse) => void;

    logout: (data: LogoutResponse & { sessionId: string }) => void;

    register: (data: RegisterResponse) => void;

    license: (data: LicenseResponse) => void;

    fetchStats: (data: FetchStatsResponse) => void;

    ban: (data: BanResponse & { sessionId: string }) => void;

    changeUsername: (
        data: ChangeUsernameResponse & { newUsername: string },
    ) => void;

    checkblacklist: (data: CheckBlacklistResponse) => void;

    check: (data: CheckResponse) => void;

    file: (data: DownloadResponse) => void;

    fetchOnline: (data: FetchOnlineUsersResponse) => void;

    forgot: (data: ForgotPasswordResponse & { username: string }) => void;

    chatget: (data: GetChatResponse) => void;

    getvar: (data: GetUserVarResponse) => void;

    log: (data: LogResponse & { msg: string; pcUser: string }) => void;

    chatsend: (
        data: SendChatResponse & { sentMsg: string; author: String },
    ) => void;

    setvar: (data: SetUserVarResponse) => void;

    upgrade: (data: UpgradeResponse) => void;

    webhook: (data: WebhookResponse) => void;

    var: (data: GlobalVarResponse) => void;

    // Added custom types
    request: (data: RequestResponse<EventType> & { type: EventType }) => void;

    response: (data: EventMap[EventType] & { type: EventType }) => void;

    metadata: (
        data: BaseResponse & {
            metaData: METADATA;
        },
    ) => void;

    error: (data: {
        success: false;
        message: string;
        errorCode: Error_Code;
        type: EventType;
    }) => void;
    instance: (data: BaseResponse) => void;

    session: (data: any) => void;
    ratelimit: (data: BaseResponse) => void;
};

type Error_Code =
    | "seesionKilled"
    | "noSessionID"
    | "notInitialized"
    | "notLoggedIn"
    | "unsupportedVarType"
    | "unknown"
    | "noChatChannel"
    | "invalidClientApi";

export enum ERROR_CODE {
    K_SK = "seesionKilled",
    K_NSID = "noSessionID",
    K_NI = "notInitialized",
    K_NLI = "notLoggedIn",
    K_UVT = "unsupportedVarType",
    K_U = "unknown",
    K_NCC = "noChatChannel",
    K_ICA = "invalidClientApi",
}

export interface RequestResponse<EType extends EventType> {
    request: {
        url: string;
        params: Record<string, any>;
    };
    response: EventMap[EType];
}

// Parameters for a webhook request
export interface Webhook {
    webId: string;
    sessionId: string;
    params?: string;
    body?: string;
    contType?: string | undefined;
}

// Parameters for a file download request
export interface Download {
    fileId: string;
    sessionId: string;
}

// Parameters for an upgrade request
export interface Upgrade {
    username: string;
    key: string;
    sessionId: string;
}

// Parameters for a log request
export interface Log {
    pcUser: string;
    msg: string;
    sessionId: string;
}

// Parameters for a password reset request
export interface ForgotPassword {
    username: string;
    email: string;
    sessionId: string;
}

// Parameters for changing a username
export interface ChangeUsername {
    newUsername: string;
    sessionId: string;
}

// Parameters for banning a user
export interface Ban {
    reason: string;
    sessionId: string;
}

// Parameters for a login request
export interface Login {
    username: string;
    password: string;
    hwid?: string;
    sessionId: string;
}

// Parameters for a logout request
export interface Logout {
    sessionId: string;
}

// Parameters for a license verification request
export interface License {
    license: string;
    sessionId: string;
}

// Parameters for user registration
export interface Register {
    username: string;
    password: string;
    key: string;
    sessionId: string;
    email?: string;
    metaData: METADATA;
}

// Parameters for fetching statistics
export interface FetchStats {
    sessionId: string;
}

// Parameters for checking blacklist
export interface CheckBlacklist {
    hwid: string;
    sessionId: string;
}

// Parameters for fetching online users
export interface FetchOnlineUsers {
    sessionId: string;
}

// Parameters for a general check
export interface Check {
    sessionId: string;
    skipResponse: boolean;
}

export interface GetMetaData {
    sessionId: string;
    skipResponse?: boolean;
    skipError?: boolean;
}

export interface SetMetaData {
    sessionId: string;
    metaData: METADATA;
    skipResponse?: boolean;
    skipError?: boolean;
}

export interface GetMetaDataResponse extends BaseResponse {
    metaData?: METADATA;
    nonce?: string;
}

export interface SetMetaDataResponse extends BaseResponse {
    nonce?: string;
}

// Params needed for making a request
export interface MakeRequestParams extends Record<string, any> {
    type: EventType;
}
// Parameters for making a request
export interface MakeRequest {
    params: MakeRequestParams;
    skipResponse?: boolean;
    skipError?: boolean;
}
export interface SendChat {
    channel: string;
    message: string;
    username: string;
    sessionId: string;
}
export interface SendChatResponse extends BaseResponse {}
export interface SendChatParams {
    type: EVENT_TYPE.CHAT_SEND;
    channel: string;
    message: string;
    sessionid: string;
}
export interface GetChat {
    channel: string;
    sessionId: string;
}
export interface Message {
    author: string;
    message: string;
    timestamp: string;
}
export interface GetChatResponse extends BaseResponse {
    messages: Message[];
}
export interface GetChatParams {
    type: EVENT_TYPE.CHAT_GET;
    channel: string;
    sessionid: string;
}

// Parameters for setting a variable
export interface SetUserVar {
    varId: string;
    varData: string;
    sessionId: string;
    skipResponse: boolean;
    skipError: boolean;
}

// Response after setting a user variable
export interface SetUserVarResponse extends BaseResponse {
    nonce: string;
}
// Parameters for getting a variable
export interface GetUserVar {
    varId: string;
    sessionId: string;
    skipResponse: boolean;
    skipError: boolean;
}

// Response after getting a user variable
export interface GetUserVarResponse extends BaseResponse {
    response: any;
    nonce: string;
}

// Parameters for setting a variable
export interface VarParams {
    type: "var" | "getvar" | "setvar";
    var: string;
    data?: string;
    sessionid: string;
}

export interface GlobalVar {
    varId: string;
    sessionId: string;
    skipResponse: boolean;
    skipError: boolean;
}
export interface GlobalVarResponse extends BaseResponse {
    nonce: string;
}

export interface Var {
    /**
     * Mange a users variables
     */
    user: {
        /**
         * Get a users variable.
         *
         * @param {GetUserVar} data - The data needed to set the metadata.
         * @param {GetUserVar['varId']} data.varId - The ID of the variable.
         * @param {GetUserVar['sessionId']} data.sessionID - The session ID from the current session.
         * @param {GetUserVar['skipError']} [data.skipError = false] - If to skip logging the error or not default false.
         * @param {GetUserVar['skipResponse']} [data.skipResponse = false] - If to skip logging the response or not default false.
         * @returns {Promise<GetUserVarResponse>} `response` - the response from getting the users variable.
         */
        get: (data: GetUserVar) => Promise<GetUserVarResponse>;
        /**
         * Set a users variable.
         *
         * @param {SetUserVar} data - The data needed to set the metadata.
         * @param {SetUserVar['varId']} data.varId - The ID of the variable.
         * @param {SetUserVar['varData']} data.varData - The variable data.
         * @param {SetUserVar['sessionId']} data.sessionID - The session ID from the current session.
         * @param {SetUserVar['skipError']} [data.skipError = false] - If to skip logging the error or not default false.
         * @param {SetUserVar['skipResponse']} [data.skipResponse = false] - If to skip logging the response or not default false.
         * @returns {Promise<SetUserVarResponse>} `response` - the response from setting the users variable.
         */
        set: (data: SetUserVar) => Promise<SetUserVarResponse>;
    };
    /**
     * Get a global variable.
     *
     * @param {GlobalVar} data - The data needed to set the metadata.
     * @param {GlobalVar['varId']} data.varId - The ID of the variable.
     * @param {GlobalVar['sessionId']} data.sessionID - The session ID from the current session.
     * @param {GlobalVar['skipError']} [data.skipError = false] - If to skip logging the error or not default false.
     * @param {GlobalVar['skipResponse']} [data.skipResponse = false] - If to skip logging the response or not default false.
     * @returns {Promise<GlobalVarResponse>} `response` - the response from getting a global variable.
     */
    get: (data: GlobalVar) => Promise<GlobalVarResponse>;
}
export interface MetaData {
    /**
     * Retrieve a users metadata from session ID.
     *
     * @param {GetMetaData} data - The data needed to get the metadata.
     * @param {GetMetaData['sessionId']} data.sessionID - The session ID from the current session.
     * @param {GetMetaData['skipError']} [data.skipError = false] - If to skip logging the error or not default false.
     * @param {GetMetaData['skipResponse']} [data.skipResponse = false] - If to skip logging the response or not default false.
     * @returns {Promise<GetMetaDataResponse>} `response` - the response from getting the users metadata.
     */
    get: (data: GetMetaData) => Promise<GetMetaDataResponse>;

    /**
     * Set a users metadata from session ID.
     *
     * @param {SetMetaData} data - The data needed to set the metadata.
     * @param {SetMetaData['sessionId']} data.sessionID - The session ID from the current session.
     * @param {SetMetaData['skipError']} [data.skipError = false] - If to skip logging the error or not default false.
     * @param {SetMetaData['skipResponse']} [data.skipResponse = false] - If to skip logging the response or not default false.
     * @returns {Promise<SetMetaDataResponse>} `response` - the response from getting the users metadata.
     */
    set: (data: SetMetaData) => Promise<SetMetaDataResponse>;
}

export interface Chat {
    /**
     * Get chat messages from a specific channel.
     *
     * @see https://keyauth.readme.io/reference/chat-channel
     *
     * @property {GetChat} `params` - The parameters for getting chat messages.
     * @param {GetChat['channel']} `params.channel` - The channel from which to retrieve chat messages.
     * @param {GetChat['sessionId']} `params.sessionId` - The session ID for the request.
     * @returns {Promise<GetChatResponse>} `GetChatResponse` - A Promise that resolves with the response containing chat messages.
     * @throws {Error} `Error` - Throws an error if API initialization is required.
     */
    get: (data: GetChat) => Promise<GetChatResponse>;
    /**
     * Send a chat message to a specific channel.
     *
     * @see https://keyauth.readme.io/reference/set-chat-channels
     *
     * @param {SendChat} `params` - The parameters for sending a chat message.
     * @param {SendChat['channel']} `params.channel` - The channel where the message will be sent.
     * @param {SendChat['message']} `params.message` - The message content to be sent.
     * @param {SendChat['username']} `params.username` - The username of the sender.
     * @param {SendChat['sessionId']} `params.sessionId` - The session ID of the sender.
     * @returns {Promise<SendChatResponse>} `SendChatResponse` - A Promise that resolves with the response of the chat message sending.
     * @throws {Error} `Error` - Throws an error if API initialization is required.
     */
    send: (data: SendChat) => Promise<SendChatResponse>;
}
