// @keyauthjs/client/dist/client/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// Import necessary modules and types
const rateLimiter_1 = require("../utils/rateLimiter");
const axios_1 = __importStar(require("axios"));
const types_1 = require("../types");
const logger_1 = __importDefault(require("../utils/logger"));
const constants_1 = require("../utils/constants");
const helpers_1 = require("../utils/helpers");
/**
 * KeyAuth Client API Wrapper
 *
 * This class provides a TypeScript wrapper for the KeyAuth.cc API.
 *
 * It allows you to interact with various API endpoints for user management and more.
 */
class Api {
    /**
     * KeyAuth Client API Wrapper
     *
     * This class provides a TypeScript wrapper for the KeyAuth.cc API.
     *
     * It allows you to interact with various API endpoints for user management and more.
     *
     * Constructor for the Api class.
     * @param {App} `app` - App details can be found on the keyauth dashboard.
     * @param {App['name']} `app.name` - The application name.
     * @param {App['ownerid']} `app.ownerid` - The application ownerID.
     * @param {App['ver']} `app.ownerid` - The application version.
     *
     * @param {ClientOptions} [options] - Client api options all optional.
     * @param {ClientOptions['baseUrl']} `options.baseUrl` - To change the base url from https://keyauth.win/api/1.2/ to anything you like.
     * @param {ClientOptions['convertTimes']} `options.convertTimes` - To convert all times to readable times example like sub expiry.
     * @param {ClientOptions['logger']} `options.logger` - All the logger options.
     * @param {ClientOptions['ratelimit']} `options.ratelimit` - override the global rate limit set on the package.
     */
    constructor(app, options) {
        var _a, _b, _c, _d;
        /**
         * Flag to track whether the client has been initialized.
         * @private
         */
        this._initializedClient = false;
        this.metaData = {
            get: async ({ sessionId, skipError = false, skipResponse = false, }) => {
                // Check if API initialization is required
                if (!this._checkInitialization()) {
                    throw "API Initialization Required: Please initialize the API first.";
                }
                // Check if the user is logged in
                if (!(await this._checkUserLogin(sessionId))) {
                    return Promise.reject();
                }
                // Get metadata from the server
                const { message, success, time, nonce, response } = await this.var.user.get({
                    sessionId,
                    varId: "metaData",
                    skipError,
                    skipResponse,
                });
                if (success) {
                    // Emit an event and return successful response
                    this._eventEmitter.emit(types_1.EVENT_TYPE.METADATA, {
                        message: "MetaData Successfully retrieved",
                        success,
                        time,
                        metaData: JSON.parse(response),
                    });
                    return {
                        message: "MetaData Successfully retrieved",
                        success,
                        time,
                        metaData: JSON.parse(response),
                    };
                }
                else {
                    // Return a response for failed metadata retrieval
                    return {
                        message: "Failed to retrieve metaData. Try setting metaData first",
                        success,
                        time,
                    };
                }
            },
            set: async ({ metaData, sessionId, skipError = false, skipResponse = false, }) => {
                // Check if API initialization is required
                if (!this._checkInitialization()) {
                    throw "API Initialization Required: Please initialize the API first.";
                }
                // Check if the user is logged in
                if (!(await this._checkUserLogin(sessionId))) {
                    return Promise.reject();
                }
                // Set metadata on the server
                const { message, nonce, success, time } = await this.var.user.set({
                    sessionId,
                    skipError,
                    skipResponse,
                    varData: JSON.stringify(metaData),
                    varId: "metaData",
                });
                // Emit an event and return successful response
                this._eventEmitter.emit(types_1.EVENT_TYPE.METADATA, {
                    message: "MetaData Successfully updated",
                    success,
                    time,
                    metaData,
                });
                return {
                    message: "MetaData Successfully updated",
                    success,
                    time,
                    nonce,
                };
            },
        };
        this.var = {
            get: async ({ sessionId, varId, skipResponse = false, skipError = false, }) => {
                // Log the GLOBAL operation
                this._logger.debug(types_1.EVENT_TYPE.VAR, "Running var.");
                // Check if the API is initialized
                if (!this._checkInitialization()) {
                    // If not initialized, reject the operation
                    return Promise.reject("API Initialization Required: Please initialize the API first.");
                }
                // Prepare GLOBAL variable parameters
                const varParams = {
                    type: types_1.EVENT_TYPE.VAR,
                    var: varId,
                    sessionid: sessionId,
                };
                // Log the GLOBAL request
                this._logger.debug(types_1.EVENT_TYPE.VAR, "Sending var request.");
                // Send the GLOBAL request and wait for the response
                const response = await this._makeRequest({
                    params: { ...varParams },
                    skipError,
                    skipResponse,
                });
                // Log that the GLOBAL operation is complete and return the response
                this._logger.debug(types_1.EVENT_TYPE.VAR, "VAR request complete, Returning response.");
                return {
                    ...response,
                };
            },
            user: {
                get: async ({ sessionId, varId, skipResponse = false, skipError = false, }) => {
                    // Log the GET operation
                    this._logger.debug(types_1.EVENT_TYPE.GET_VAR, "Running get user var.");
                    // Check if the API is initialized
                    if (!this._checkInitialization()) {
                        // If not initialized, reject the operation
                        return Promise.reject("API Initialization Required: Please initialize the API first.");
                    }
                    // Prepare GET variable parameters
                    const getVarParams = {
                        type: types_1.EVENT_TYPE.GET_VAR,
                        sessionid: sessionId,
                        var: varId,
                    };
                    // Log the GET request
                    this._logger.debug(types_1.EVENT_TYPE.GET_VAR, "Sending get user var request.");
                    // Send the GET request and wait for the response
                    const response = await this._makeRequest({
                        params: { ...getVarParams },
                        skipResponse,
                        skipError,
                    });
                    // Log that the GET operation is complete and return the response
                    this._logger.debug(types_1.EVENT_TYPE.GET_VAR, "Get user var request complete, Returning response.");
                    return { ...response };
                },
                set: async ({ sessionId, varData, varId, skipResponse = false, skipError = false, }) => {
                    // Log the SET operation
                    this._logger.debug(types_1.EVENT_TYPE.SET_VAR, "Running Set user var.");
                    // Check if the API is initialized
                    if (!this._checkInitialization()) {
                        // If not initialized, reject the operation
                        return Promise.reject("API Initialization Required: Please initialize the API first.");
                    }
                    // Prepare SET variable parameters
                    const setVarParams = {
                        type: types_1.EVENT_TYPE.SET_VAR,
                        sessionid: sessionId,
                        data: varData,
                        var: varId,
                    };
                    // Log the SET request
                    this._logger.debug(types_1.EVENT_TYPE.SET_VAR, "Sending set user var request.");
                    // Send the SET request and wait for the response
                    const response = await this._makeRequest({
                        params: { ...setVarParams },
                        skipResponse,
                        skipError,
                    });
                    // Log that the SET operation is complete and return the response
                    this._logger.debug(types_1.EVENT_TYPE.SET_VAR, "Set user var request complete, Returning response.");
                    return { ...response };
                },
            },
        };
        this.chat = {
            get: async ({ channel, sessionId, }) => {
                this._logger.debug(types_1.EVENT_TYPE.CHAT_GET, "Running get chat.");
                // Check if API initialization is required
                if (!this._checkInitialization()) {
                    throw "API Initialization Required: Please initialize the API first.";
                }
                // Construct the get chat parameters
                const getChatParams = {
                    type: types_1.EVENT_TYPE.CHAT_GET,
                    sessionid: sessionId,
                    channel,
                };
                this._logger.debug(types_1.EVENT_TYPE.CHAT_GET, "Sending get chat request.");
                // Make the API request to retrieve chat messages
                const response = await this._makeRequest({
                    params: { ...getChatParams },
                });
                // Emit a 'chatget' event if chat messages were successfully retrieved
                if (response.success) {
                    this._eventEmitter.emit("chatget", {
                        ...response,
                    });
                }
                this._logger.debug(types_1.EVENT_TYPE.CHAT_GET, "Get chat request complete, Returning response.");
                return { ...response };
            },
            send: async ({ channel, message, username, sessionId, }) => {
                this._logger.debug(types_1.EVENT_TYPE.CHAT_SEND, "Running set chat.");
                // Check if API initialization is required
                if (!this._checkInitialization()) {
                    throw "API Initialization Required: Please initialize the API first.";
                }
                // Construct the send chat parameters
                const sendChatParams = {
                    type: types_1.EVENT_TYPE.CHAT_SEND,
                    sessionid: sessionId,
                    channel,
                    message,
                };
                this._logger.debug(types_1.EVENT_TYPE.CHAT_SEND, "Sending set chat request.");
                // Make the API request to send the chat message
                const response = await this._makeRequest({
                    params: { ...sendChatParams },
                });
                // Emit a 'chatsend' event if the chat message was sent successfully
                if (response.success) {
                    this._eventEmitter.emit("chatsend", {
                        author: username,
                        sentMsg: message,
                        ...response,
                    });
                }
                this._logger.debug(types_1.EVENT_TYPE.CHAT_SEND, "Set chat request complete, Returning response.");
                return { ...response };
            },
        };
        // Initialize event emitter
        const eventEmitter = new types_1.KeyauthEventEmitter();
        this._eventEmitter = eventEmitter;
        // Initialize logger with provided options or empty object
        this._logger = new logger_1.default((_a = { ...options === null || options === void 0 ? void 0 : options.logger, name: "Keyauth API" }) !== null && _a !== void 0 ? _a : {});
        // Log custom base URL usage if provided
        if (options === null || options === void 0 ? void 0 : options.baseUrl) {
            this._logger.info(types_1.EVENT_TYPE.INSTANCE, "Using custom base url");
        }
        // Create Axios instance with base URL and headers
        this._axiosInstance = axios_1.default.create({
            baseURL: (_b = options === null || options === void 0 ? void 0 : options.baseUrl) !== null && _b !== void 0 ? _b : constants_1.BASE_URL,
            headers: constants_1.HEADERS,
        });
        // Initialize rate limiter with default or provided options
        this._rateLimiter = new rateLimiter_1.RateLimiter((_c = options === null || options === void 0 ? void 0 : options.ratelimit) !== null && _c !== void 0 ? _c : { maxTokens: 10, refillRate: 5000 });
        // Determine whether to convert timestamps to local dates
        this._convertTimes = (_d = options === null || options === void 0 ? void 0 : options.convertTimes) !== null && _d !== void 0 ? _d : false;
        // Store information about the app
        this._app = app;
        // Log Axios instance creation
        this._logger.debug(types_1.EVENT_TYPE.INSTANCE, "Keyauth instance created.");
    }
    /**
     * Register an event listener for a specific event type.
     *
     * @template E - The type of the event.
     * @param {E} event - The event type to listen for.
     * @param {EventMap[E]} callback - The callback function to execute when the event occurs.
     */
    on(event, callback) {
        this._eventEmitter.on(event, callback);
    }
    /**
     * Register a one-time event listener for a specific event type.
     * The listener will be automatically removed after it's been called once.
     *
     * @template E - The type of the event.
     * @param {E} event - The event type to listen for.
     * @param {EventMap[E]} callback - The callback function to execute when the event occurs.
     */
    once(event, callback) {
        this._eventEmitter.once(event, callback);
    }
    // Private functions
    // ------------------------------------------
    /**
     * Makes a request to the Keyauth API with the provided parameters.
     *
     * @param {MakerRequest} options - The parameters for the API request.
     * @returns {Promise<ApiResponse>} A promise that resolves to the API response.
     * @throws {AxiosError} If there is an issue with the Axios HTTP request.
     * @private
     */
    async _makeRequest({ params, skipError = false, skipResponse = false, }) {
        // Capture the start time to measure request duration
        const startTime = Date.now();
        try {
            this._logger.debug(types_1.EVENT_TYPE.REQUEST, "Making a request to keyauth API.");
            // Check if the rate limiter has hit its rate limit
            if (this._rateLimiter.hasHitRateLimit()) {
                this._eventEmitter.emit(types_1.EVENT_TYPE.RESPONSE, {
                    type: types_1.EVENT_TYPE.RATE_LIMIT,
                    success: false,
                    message: `Client Api rate limit hit please wait ${this._rateLimiter.getTimeUntilCanMakeRequestString()}`,
                    time: startTime - Date.now(),
                });
                this._logger.debug(types_1.EVENT_TYPE.REQUEST, `Rate limit hit please wait ${this._rateLimiter.getTimeUntilCanMakeRequestString()}`);
                // Wait until it's possible to make a request again
                await this._rateLimiter.waitUntilCanMakeRequest();
                // Retry the request after the rate limit is reset
                return this._makeRequest({ params, skipResponse, skipError });
            }
            // Perform the Axios HTTP request to the Keyauth API
            const response = await this._axiosInstance.request({
                params: {
                    ...params,
                    name: this._app.name,
                    ownerid: this._app.ownerid,
                },
            });
            // Calculate the time taken for the request
            const endTime = Date.now();
            const responseTime = endTime - startTime;
            // Handle different types of API responses
            if (response.data === "KeyAuth_Invalid") {
                // Handle the case where the API client is not set up correctly
                response.data = {
                    success: false,
                    message: "Keyauth API client not set up correctly!",
                    time: responseTime,
                };
            }
            if (params.type === types_1.EVENT_TYPE.LOG && response.status === 200) {
                // Handle a successful log request
                response.data = {
                    success: true,
                    message: "Log successfully sent.",
                };
            }
            if (params.type === types_1.EVENT_TYPE.WEBHOOK && response.status === 200) {
                // Handle a successful webhook request
                response.data = {
                    success: response.data.success,
                    message: response.data.message,
                    nonce: response.data.nonce,
                    data: response.data.response,
                };
            }
            // Emit an event for the API response (if not skipped)
            if (!skipResponse) {
                this._eventEmitter.emit(types_1.EVENT_TYPE.RESPONSE, {
                    ...response.data,
                    time: responseTime,
                    type: params.type,
                });
            }
            // Emit an event for the API request and return the response
            this._eventEmitter.emit(types_1.EVENT_TYPE.REQUEST, {
                type: params.type,
                request: {
                    url: this._axiosInstance.defaults.baseURL,
                    params,
                },
                response: {
                    ...response.data,
                    time: responseTime,
                },
            });
            if (response.data.success === false) {
                if (skipError)
                    return;
                if (params.sessionid === "" &&
                    String(response.data.message).startsWith("Session not found.")) {
                    this._eventEmitter.emit(types_1.EVENT_TYPE.ERROR, {
                        type: params.type,
                        success: false,
                        message: response.data.message,
                        errorCode: types_1.ERROR_CODE.K_NSID,
                    });
                }
                else if (String(response.data.message).startsWith("Session not found.")) {
                    this._eventEmitter.emit(types_1.EVENT_TYPE.ERROR, {
                        type: params.type,
                        success: false,
                        message: "The session was killed!",
                        errorCode: types_1.ERROR_CODE.K_SK,
                    });
                }
                else if (String(response.data.message) === "Chat channel not found") {
                    this._eventEmitter.emit(types_1.EVENT_TYPE.ERROR, {
                        type: params.type,
                        success: false,
                        message: "Chat channel not found",
                        errorCode: types_1.ERROR_CODE.K_NCC,
                    });
                }
                else if (String(response.data.message).startsWith("Keyauth API client")) {
                    this._eventEmitter.emit(types_1.EVENT_TYPE.ERROR, {
                        type: params.type,
                        success: false,
                        message: response.data.message,
                        errorCode: types_1.ERROR_CODE.K_ICA,
                    });
                }
                else {
                    this._eventEmitter.emit(types_1.EVENT_TYPE.ERROR, {
                        type: params.type,
                        success: false,
                        message: response.data.message,
                        errorCode: types_1.ERROR_CODE.K_U,
                    });
                }
            }
            return {
                ...response.data,
                time: responseTime,
            };
        }
        catch (error) {
            if (error instanceof axios_1.AxiosError) {
                // Handle Axios HTTP request errors
                this._logger.error(types_1.EVENT_TYPE.ERROR, error.message);
                this._eventEmitter.emit(types_1.EVENT_TYPE.ERROR, {
                    type: params.type,
                    success: false,
                    message: error.message,
                    errorCode: types_1.ERROR_CODE.K_U,
                });
            }
            else {
                // Handle other types of errors
                this._logger.error(types_1.EVENT_TYPE.ERROR, `${error}`);
                this._eventEmitter.emit(types_1.EVENT_TYPE.ERROR, {
                    type: params.type,
                    success: false,
                    message: error.message,
                    errorCode: types_1.ERROR_CODE.K_U,
                });
            }
            // TODO Add more specific error handling if needed
        }
    }
    /**
     * Checks if the API client has been properly initialized with required configurations.
     *
     * @returns {boolean} True if the API client is initialized, false otherwise.
     * @throws {Error} Throws an error if the API client is not properly initialized.
     * @private
     */
    _checkInitialization() {
        if (!this._app) {
            throw new Error("API client is not properly initialized. Missing 'app' configuration.");
        }
        if (!this._axiosInstance) {
            throw new Error("API client is not properly initialized. Missing 'axiosInstance' configuration.");
        }
        if (!this._rateLimiter) {
            throw new Error("API client is not properly initialized. Missing 'rateLimiter' configuration.");
        }
        if (!this._logger) {
            throw new Error("API client is not properly initialized. Missing 'logger' configuration.");
        }
        if (!this._initializedClient) {
            this._logger.error(types_1.EVENT_TYPE.INIT, "API Initialization Required: Please initialize the API first.");
            this._eventEmitter.emit(types_1.EVENT_TYPE.ERROR, {
                type: types_1.EVENT_TYPE.INIT,
                success: false,
                message: "API Initialization Required: Please initialize the API first.",
                errorCode: types_1.ERROR_CODE.K_NI,
            });
        }
        return true;
    }
    /**
     * Checks if a user is logged in based on their session ID.
     *
     * @param {string} sessionId - The session ID of the user to check.
     * @returns {Promise<void>} Resolves if the user is logged in, rejects with an error otherwise.
     * @throws {Error} Throws an error if the user is not logged in or if the API client is not initialized.
     * @private
     */
    async _checkUserLogin(sessionId) {
        // Ensure that the API client is properly initialized
        if (!this._checkInitialization())
            return false;
        // Check if the user is logged in by making a request to keyauth api
        const { success } = await this.check({ sessionId, skipResponse: true });
        if (success)
            return true;
        this._logger.error(types_1.EVENT_TYPE.ERROR, "User Not Logged In: Please login before accessing user data.");
        this._eventEmitter.emit(types_1.EVENT_TYPE.ERROR, {
            type: types_1.EVENT_TYPE.LOG_IN,
            success: false,
            message: "User Not Logged In: Please login before accessing user data.",
            errorCode: types_1.ERROR_CODE.K_NLI,
        });
        return false;
    }
    // ------------------------------------------
    // User Authentication and Management Methods
    // ------------------------------------------
    /**
     * Initialize the KeyAuth API.
     * This function initializes the API client, ensuring it's ready for use.
     *
     * @see https://keyauth.readme.io/reference/initialization
     * @returns {Promise<InitResponse>} - A promise containing the initialization response.
     */
    async init() {
        // Capture the start time to measure execution time
        const startTime = Date.now();
        // Log the initialization process
        this._logger.debug(types_1.EVENT_TYPE.INIT, "Initializing the API.");
        // Check if the API is already initialized
        if (this._initializedClient) {
            this._logger.debug(types_1.EVENT_TYPE.INIT, "API is already initialized.");
            // Return a success response with a message and execution time
            return {
                success: true,
                message: "Already initialized",
                time: startTime - Date.now(),
            };
        }
        // Prepare initialization parameters
        const initParams = {
            type: types_1.EVENT_TYPE.INIT,
            ver: this._app.ver,
            name: this._app.name,
            ownerid: this._app.ownerid,
        };
        // Log the initialization request
        this._logger.debug(types_1.EVENT_TYPE.INIT, "Sending initialization request.");
        // Send the initialization request and wait for the response
        const response = await this._makeRequest({ params: { ...initParams } });
        // Update the internal state to mark the API as initialized
        this._initializedClient = true;
        // Emit an event to indicate the completion of initialization
        this._eventEmitter.emit(types_1.EVENT_TYPE.INIT, { ...response });
        // Log that initialization is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.INIT, "Initialization complete. Returning response.");
        return {
            ...response,
        };
    }
    /**
     * Log in a user with provided credentials.
     * This function attempts to authenticate a user with a username and password.
     *
     * @see https://keyauth.readme.io/reference/login
     *
     * @template D - A generic parameter representing metaData.
     * @param {Login} `options` - The login options including username, password, and session information.
     * @param {Login['sessionId']} `options.sessionId` - The current session ID.
     * @param {Login['username']} `options.username` - The username of the person logging in.
     * @param {Login['password']} `options.password` - The password of the person logging in.
     * @param {Login['hwid']} [options.hwid] - The hwid of the person logging in if needed.
     * @returns {Promise<LoginResponse<D>>} - A promise containing the login response, including user metadata.
     */
    async login({ username, password, hwid, sessionId, }) {
        // Log the login process
        this._logger.debug(types_1.EVENT_TYPE.LOG_IN, "Logging in user.");
        // Check if the API is initialized
        if (!this._checkInitialization())
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        // Prepare login parameters
        const loginParams = {
            type: types_1.EVENT_TYPE.LOG_IN,
            pass: password,
            sessionid: sessionId,
            username,
            hwid,
        };
        // Log the login request
        this._logger.debug(types_1.EVENT_TYPE.LOG_IN, "Sending login request.");
        // Send the login request and wait for the response
        const response = await this._makeRequest({
            params: { ...loginParams },
        });
        // Check if the login was successful
        if (response === null || response === void 0 ? void 0 : response.success) {
            // If requested & convert timestamps [true], convert timestamps in the response to local dates
            if (this._convertTimes) {
                response.info = (0, helpers_1.convertTimestampsToLocalDates)(response.info);
            }
            // Fetch user metadata
            const { metaData } = await this.metaData.get({
                sessionId,
                skipResponse: true,
                skipError: true,
            });
            // Emit an event to indicate a successful login
            if (response.info) {
                this._eventEmitter.emit(types_1.EVENT_TYPE.LOG_IN, {
                    ...response,
                    metaData: metaData,
                });
            }
            // Log that the login is complete and return the response with user metadata
            this._logger.debug(types_1.EVENT_TYPE.LOG_IN, "Login complete. Returning response.");
            return {
                ...response,
                metaData: metaData,
            };
        }
        else {
            // Log that the login is complete and return the response
            this._logger.debug(types_1.EVENT_TYPE.LOG_IN, "Login complete. Returning response.");
            return {
                ...response,
            };
        }
    }
    /**
     * Log out a user with the provided session ID.
     * This function logs out a user, revoking their session and clearing their user information.
     *
     * @see https://keyauth.readme.io/reference/logout
     *
     * @param {Logout} `options` - The logout options including the session ID.
     * @param {Logout['sessionId']} `options.sessionId` - The current session ID.
     * @returns {Promise<LogoutResponse>} - A promise containing the logout response.
     */
    async logout({ sessionId }) {
        // Log the logout process
        this._logger.debug(types_1.EVENT_TYPE.LOG_OUT, "Logging out user.");
        // Check if the user is currently logged in
        if (!(await this._checkUserLogin(sessionId))) {
            // If not logged in, reject the logout request
            return Promise.reject();
        }
        // Prepare logout parameters
        const logoutParams = {
            type: types_1.EVENT_TYPE.LOG_OUT,
            sessionid: sessionId,
        };
        // Log the logout request
        this._logger.debug(types_1.EVENT_TYPE.LOG_OUT, "Sending logout request.");
        // Send the logout request and wait for the response
        const response = await this._makeRequest({
            params: { ...logoutParams },
        });
        // Clear user info and update nonce after successful logout
        this._logger.debug(types_1.EVENT_TYPE.LOG_OUT, "Clearing user info and updating nonce.");
        // Emit an event to indicate a successful logout
        this._eventEmitter.emit(types_1.EVENT_TYPE.LOG_OUT, {
            ...response,
            sessionId,
        });
        // Log that the logout is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.LOG_OUT, "Logout complete. Returning response.");
        return { ...response };
    }
    /**
     * Register a new user with the provided registration details.
     * This function creates a new user account and associates it with the provided session ID.
     *
     * @see https://keyauth.readme.io/reference/register
     *
     * @template D - A generic parameter representing metaData.
     * @param {Register} `options` - The registration options including username, password, key, etc.
     * @param {Register<D>['username']} `options.username` - The username of the new user.
     * @param {Register<D>['password']} `options.password` - The password of the new user.
     * @param {Register<D>['key']} `options.key` - The key needed for the user to register.
     * @param {Register<D>['sessionId']} `options.sessionId` - The session Id of the current session.
     * @param {Register<D>['metaData']} `options.metaData` - The metaData to set upon registering.
     * @param {Register<D>['email']} [options.email] - The email address of the new user if you would to set one.
     * @returns {Promise<RegisterResponse>} - A promise containing the registration response.
     */
    async register({ username, password, key, sessionId, email, metaData, }) {
        // Log the registration process
        this._logger.debug(types_1.EVENT_TYPE.REGISTER, "Registering a new user.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the registration request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Prepare registration parameters
        const registerParams = {
            type: types_1.EVENT_TYPE.REGISTER,
            username,
            pass: password,
            key,
            sessionid: sessionId,
            email,
        };
        // Log the registration request
        this._logger.debug(types_1.EVENT_TYPE.REGISTER, "Sending registration request.");
        // Send the registration request and wait for the response
        const response = await this._makeRequest({
            params: { ...registerParams },
        });
        // Log and emit an event if user info is available
        this._logger.debug(types_1.EVENT_TYPE.REGISTER, "Updating user info if available.");
        if (response === null || response === void 0 ? void 0 : response.info) {
            this._eventEmitter.emit(types_1.EVENT_TYPE.REGISTER, { ...response.data });
        }
        // Log that registration is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.REGISTER, "Registration complete. Returning response.");
        // If metadata is provided, set it for the user
        if (metaData) {
            await this.metaData.set({
                sessionId,
                metaData,
                skipResponse: true,
            });
        }
        return {
            ...response,
        };
    }
    /**
     * Log in or register a user using a license key.
     * This function allows a user to authenticate or register with the provided license key.
     *
     * @see https://keyauth.readme.io/reference/license
     *
     * @param {License} options - The license options including the license key, session ID, etc.
     * @param {License['license']} `options.license` - The license key needed to register.
     * @param {License['sessionId']} `options.sessionId` - The session ID of the current session.
     * @returns {Promise<LicenseResponse>} - A promise containing the license response.
     */
    async license({ license, sessionId }) {
        // Log the license process
        this._logger.debug(types_1.EVENT_TYPE.LICENSE, "Logging in or registering a user via license.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the license request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Prepare license parameters
        const licenseParams = {
            type: types_1.EVENT_TYPE.LICENSE,
            key: license,
            sessionid: sessionId,
        };
        // Log the license request
        this._logger.debug(types_1.EVENT_TYPE.LICENSE, "Sending license request.");
        // Send the license request and wait for the response
        const response = await this._makeRequest({
            params: { ...licenseParams },
        });
        // Log and emit an event if user info is available
        this._logger.debug(types_1.EVENT_TYPE.LICENSE, "Updating user info if available.");
        if (response.info) {
            this._eventEmitter.emit(types_1.EVENT_TYPE.LICENSE, { ...response.data });
        }
        // Log that the license request is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.LICENSE, "License request complete. Returning response.");
        return { ...response };
    }
    /**
     * Ban a user with a given reason.
     * This function initiates a user ban and sends a ban request to the API.
     *
     * @see https://keyauth.readme.io/reference/ban
     *
     * @param {Ban} `options` - The ban options including the ban reason, session ID, etc.
     * @param {Ban['sessionId']} `options.sessionId` - The current session ID.
     * @param {Ban['reason']} `options.reason` - The reason for banning the current user.
     * @returns {Promise<BanResponse>} - A promise containing the ban response.
     */
    async ban({ reason, sessionId }) {
        // Log the user ban initiation
        this._logger.debug(types_1.EVENT_TYPE.BAN, "Initiating user ban.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the ban request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Ensure the user is logged in by checking the session ID
        await this._checkUserLogin(sessionId);
        // Prepare ban parameters
        const banParams = {
            type: types_1.EVENT_TYPE.BAN,
            sessionid: sessionId,
            reason,
        };
        // Log the ban request
        this._logger.debug(types_1.EVENT_TYPE.BAN, "Sending ban request.");
        // Send the ban request and wait for the response
        const response = await this._makeRequest({
            params: { ...banParams },
        });
        // Log and emit a ban event with relevant information
        this._logger.debug(types_1.EVENT_TYPE.BAN, "Clearing user info.");
        this._eventEmitter.emit(types_1.EVENT_TYPE.BAN, {
            ...response,
            sessionId,
        });
        // Log that the ban request is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.BAN, "Ban request complete. Returning response.");
        return { ...response };
    }
    /**
     * Check if the current session is validated.
     *
     * This function sends a request to the API to check if the provided session ID is validated.
     *
     * @see https://keyauth.readme.io/reference/check-session
     *
     * @param {Check} `options` - The check options including the session ID and skipResponse flag.
     * @param {Check['sessionId']} `options.sessionID` - The current session ID.
     * @param {Check['skipResponse']} `options.skipResponse` - This will just not trigger the response event. [Not really needed]
     * @returns {Promise<CheckResponse>} - A promise containing the check response.
     */
    async check({ sessionId, skipResponse = false, }) {
        // Log the check operation
        this._logger.debug(types_1.EVENT_TYPE.CHECK, "Checking if current session is validated.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the check request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Prepare check parameters
        const checkParams = {
            type: types_1.EVENT_TYPE.CHECK,
            sessionid: sessionId,
        };
        // Log the check request
        this._logger.debug(types_1.EVENT_TYPE.CHECK, "Sending check request.");
        // Send the check request and wait for the response
        const response = await this._makeRequest({
            params: { ...checkParams },
            skipResponse,
        });
        // Log that the check operation is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.CHECK, "Check complete, Returning response.");
        return { ...response };
    }
    /**
     * Check if a user is blacklisted based on their hardware ID.
     *
     * This function sends a request to the API to check if the user with the provided hardware ID and session ID is blacklisted.
     *
     * @see https://keyauth.readme.io/reference/check-blacklist
     *
     * @param {CheckBlacklist} options - The check blacklist options including the hardware ID and session ID.
     * @param {CheckBlacklist['hwid']} options.hwid - The HWID of the user you would like to check.
     * @param {CheckBlacklist['sessionId']} options.sessionId - The current session ID.
     * @returns {Promise<CheckBlacklistResponse>} - A promise containing the check blacklist response.
     */
    async checkBlacklist({ hwid, sessionId, }) {
        // Log the check blacklist operation
        this._logger.debug(types_1.EVENT_TYPE.CHECK_BLACKLIST, "Checking if user is blacklisted.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the check blacklist request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Check if the user is logged in by waiting for the login check
        await this._checkUserLogin(sessionId);
        // Prepare check blacklist parameters
        const checkBlacklistParams = {
            type: types_1.EVENT_TYPE.CHECK_BLACKLIST,
            hwid,
            sessionid: sessionId,
        };
        // Log the check blacklist request
        this._logger.debug(types_1.EVENT_TYPE.CHECK_BLACKLIST, "Sending check blacklist request.");
        // Send the check blacklist request and wait for the response
        const response = await this._makeRequest({
            params: { ...checkBlacklistParams },
        });
        // Log that the check blacklist operation is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.CHECK_BLACKLIST, "Check blacklist complete, Returning response.");
        return { ...response };
    }
    /**
     * Change the username of a logged-in user.
     *
     * This function sends a request to the API to change the username of a user with the provided session ID and then logs them out.
     *
     * You user must log back in after!
     *
     * @see https://keyauth.readme.io/reference/change-username
     *
     * @param {ChangeUsername} options - The change username options including the new username and session ID.
     * @param {ChangeUsername['newUsername']} options.newUsername - The new username.
     * @param {ChangeUsername['sessionId']} options.sessionId - The current session ID.
     * @returns {Promise<ChangeUsernameResponse>} - A promise containing the change username response.
     */
    async changeUsername({ newUsername, sessionId, }) {
        // Log the change username operation
        this._logger.debug(types_1.EVENT_TYPE.CHANGE_USERNAME, "Changing users username.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the change username request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Check if the user is logged in by waiting for the login check
        if (!(await this._checkUserLogin(sessionId))) {
            // If not logged in, reject the change username request
            return Promise.reject();
        }
        // Prepare change username parameters
        const changeUsernameParams = {
            type: types_1.EVENT_TYPE.CHANGE_USERNAME,
            newUsername,
            sessionid: sessionId,
        };
        // Log the change username request
        this._logger.debug(types_1.EVENT_TYPE.CHANGE_USERNAME, "Sending change username request.");
        // Send the change username request and wait for the response
        const response = await this._makeRequest({
            params: { ...changeUsernameParams },
        });
        // Emit an event indicating the username change
        this._eventEmitter.emit(types_1.EVENT_TYPE.CHANGE_USERNAME, {
            newUsername,
            ...response,
        });
        // Log that the change username operation is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.CHANGE_USERNAME, "Change username complete, Returning response.");
        return { ...response };
    }
    /**
     * Initiate a password reset request for a user.
     *
     * This function sends a request to the API to initiate a password reset for a user with the provided email and username.
     *
     * @see https://keyauth.readme.io/reference/forgot-password
     *
     * @param {ForgotPassword} options - The forgot password options including the email, username, and session ID.
     * @param {ForgotPassword['email']} options.email - The email address of the user requesting the password reset.
     * @param {ForgotPassword['sessionId']} options.sessionId - The current session ID.
     * @param {ForgotPassword['username']} options.username - The current users username.
     * @returns {Promise<ForgotPasswordResponse>} - A promise containing the forgot password response.
     */
    async forgotPassword({ email, username, sessionId, }) {
        // Log the forgot password operation
        this._logger.debug(types_1.EVENT_TYPE.FORGOT_PASSWORD, "Running forgot password for user.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the forgot password request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Prepare forgot password parameters
        const forgotPasswordParams = {
            type: types_1.EVENT_TYPE.FORGOT_PASSWORD,
            sessionid: sessionId,
            email,
            username,
        };
        // Log the forgot password request
        this._logger.debug(types_1.EVENT_TYPE.FORGOT_PASSWORD, "Sending forgot password request.");
        // Send the forgot password request and wait for the response
        const response = await this._makeRequest({
            params: { ...forgotPasswordParams },
        });
        // Log that the forgot password operation is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.FORGOT_PASSWORD, "Forgot password request complete, Returning response.");
        // Emit an event indicating the forgot password request
        this._eventEmitter.emit(types_1.EVENT_TYPE.FORGOT_PASSWORD, {
            username,
            ...response,
        });
        return { ...response };
    }
    /**
     * Upgrade a user's account using a key.
     *
     * This function sends a request to the API to upgrade a user's account using the provided key, username, session ID, and metadata.
     *
     * @see https://keyauth.readme.io/reference/upgrade
     *
     * @param {Upgrade} options - The upgrade options including the key, username, session ID.
     * @param {Upgrade['key']} options.key - The license key they are upgrading to.
     * @param {Upgrade['username']} options.username - The current users username.
     * @param {Upgrade['sessionId']} options.sessionId - The current session ID.
     * @returns {Promise<UpgradeResponse>} - A promise containing the upgrade response.
     */
    async upgrade({ key, username, sessionId, }) {
        // Log the upgrade operation
        this._logger.debug(types_1.EVENT_TYPE.UPGRADE, "Running upgrade on user.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the upgrade request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Prepare upgrade parameters
        const upgradeParams = {
            type: types_1.EVENT_TYPE.UPGRADE,
            sessionid: sessionId,
            username,
            key,
        };
        // Log the upgrade request
        this._logger.debug(types_1.EVENT_TYPE.UPGRADE, "Sending upgrade request.");
        // Send the upgrade request and wait for the response
        const response = await this._makeRequest({
            params: { ...upgradeParams },
        });
        // Log that the upgrade operation is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.UPGRADE, "Upgrade request complete, Returning response.");
        // Emit an event indicating the user's account upgrade
        this._eventEmitter.emit(types_1.EVENT_TYPE.UPGRADE, {
            ...response,
        });
        return { ...response };
    }
    // --------------------------------------
    // Other API Methods (Webhooks, etc.)
    // --------------------------------------
    /**
     * Fetch a list of all currently online users.
     *
     * This function sends a request to the API to retrieve a list of all users who are currently online.
     *
     * @see https://keyauth.readme.io/reference/fetch-online-users
     *
     * @param {FetchOnlineUsers} options - The fetch online users options including the session ID.
     * @param {FetchOnlineUsers['sessionId']} options.sessionID - The current session ID.
     * @returns {Promise<FetchOnlineUsersResponse>} - A promise containing the response with the list of online users.
     */
    async fetchOnlineUsers({ sessionId, }) {
        var _a;
        // Log the fetch online users operation
        this._logger.debug(types_1.EVENT_TYPE.FETCH_ONLINE, "Fetching all current online users.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the fetch online users request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Prepare fetch online users parameters
        const fetchOnlineUsersParams = {
            type: types_1.EVENT_TYPE.FETCH_ONLINE,
            sessionid: sessionId,
        };
        // Log the fetch online users request
        this._logger.debug(types_1.EVENT_TYPE.FETCH_ONLINE, "Sending fetch online users request.");
        // Send the fetch online users request and wait for the response
        const response = await this._makeRequest({
            params: { ...fetchOnlineUsersParams },
        });
        // Emit an event indicating the successful fetch of online users
        this._eventEmitter.emit(types_1.EVENT_TYPE.FETCH_ONLINE, {
            ...response,
        });
        // Log that the fetch online users operation is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.FETCH_ONLINE, "Fetch online users complete, Returning response.");
        // Include the count of online users in the response
        return {
            ...response,
            count: (_a = response.users.length) !== null && _a !== void 0 ? _a : 0,
        };
    }
    /**
     * Fetch application statistics and information.
     *
     * This function sends a request to the API to retrieve statistics and information about the application.
     *
     * @see https://keyauth.readme.io/reference/fetch-stats
     *
     * @param {FetchStats} options - The fetch stats options including the session ID.
     * @param {FetchStats['sessionId']} options.sessionID - The current session ID.
     * @returns {Promise<FetchStatsResponse>} - A promise containing the response with application statistics and information.
     */
    async fetchStats({ sessionId }) {
        // Log the fetch stats operation
        this._logger.debug(types_1.EVENT_TYPE.FETCH_STATS, "Fetching application information.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the fetch stats request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Prepare fetch stats parameters
        const fetchStatsParams = {
            type: types_1.EVENT_TYPE.FETCH_STATS,
            sessionid: sessionId,
        };
        // Log the fetch stats request
        this._logger.debug(types_1.EVENT_TYPE.FETCH_STATS, "Sending fetchStats request.");
        // Send the fetch stats request and wait for the response
        const response = await this._makeRequest({
            params: { ...fetchStatsParams },
        });
        // Emit an event indicating the successful fetch of stats
        this._eventEmitter.emit(types_1.EVENT_TYPE.FETCH_STATS, { ...response });
        // Log that the fetch stats operation is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.FETCH_STATS, "Fetchstats request complete. Returning response.");
        return { ...response };
    }
    /**
     * Send a log message to the API.
     *
     * This function sends a log message to the API for tracking purposes.
     *
     * @param {Log} options - The log options including the log message, PC user flag, session ID, and passed meta data.
     * @param {Log['msg']} options.msg - The log message you would like to send.
     * @param {Log['pcUser']} options.pcUser - The pcs username or anything else you would like to set it to.
     * @param {Log['sessionId']} options.sessionId - The current session ID.
     * @returns {Promise<LogResponse>} logResponse - A promise containing the response indicating the success of the log operation.
     */
    async log({ msg, pcUser, sessionId }) {
        // Log the log operation
        this._logger.debug(types_1.EVENT_TYPE.LOG, "Running log.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the log request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Prepare log parameters
        const logParams = {
            type: types_1.EVENT_TYPE.LOG,
            sessionid: sessionId,
            message: msg,
            pcuser: pcUser,
        };
        // Log the log request
        this._logger.debug(types_1.EVENT_TYPE.LOG, "Sending log request.");
        // Send the log request and wait for the response
        const response = await this._makeRequest({
            params: { ...logParams },
        });
        // Emit an event indicating the successful log operation
        this._eventEmitter.emit(types_1.EVENT_TYPE.LOG, {
            msg,
            pcUser,
            ...response,
        });
        // Log that the log operation is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.LOG, "Log request complete, Returning response.");
        return { ...response };
    }
    /**
     * Send a webhook request to the API.
     *
     * This function sends a webhook request to the API with the specified parameters.
     *
     * @see https://keyauth.readme.io/reference/webhook
     *
     * @param {Webhook} options - The webhook options including the web ID, content type, parameters, and session ID.
     * @param {Webhook['webId']} options.webId - The ID of the webhook.
     * @param {Webhook['sessionId']} options.sessionId - The current session ID.
     * @param {Webhook['params']} options.params - The optional webhook params.
     * @param {Webhook['body']} options.body - The webhook body.
     * @param {Webhook['contType']} options.contType - The webhook content type default [Application/json].
     * @returns {Promise<WebhookResponse>} - A promise containing the response indicating the success of the webhook request.
     */
    async webhook({ webId, contType, params, sessionId, body, }) {
        // Log the webhook operation
        this._logger.debug(types_1.EVENT_TYPE.WEBHOOK, "Running webhook on user.");
        // Check if the API is initialized
        if (!this._checkInitialization()) {
            // If not initialized, reject the webhook request
            return Promise.reject("API Initialization Required: Please initialize the API first.");
        }
        // Prepare webhook parameters
        const webhookParams = {
            type: types_1.EVENT_TYPE.WEBHOOK,
            sessionid: sessionId,
            webid: webId,
            params,
            body,
            conttype: contType,
        };
        // Log the webhook request
        this._logger.debug(types_1.EVENT_TYPE.WEBHOOK, "Sending webhook request.");
        // Send the webhook request and wait for the response
        const response = await this._makeRequest({
            params: { ...webhookParams },
        });
        // Emit an event indicating the successful webhook operation
        this._eventEmitter.emit(types_1.EVENT_TYPE.WEBHOOK, {
            ...response,
        });
        // Log that the webhook operation is complete and return the response
        this._logger.debug(types_1.EVENT_TYPE.WEBHOOK, "Webhook request complete, Returning response.");
        return { ...response };
    }
    /**
     * Downloads a file with the specified file ID using the provided session ID.
     *
     * @see https://keyauth.readme.io/reference/download
     *
     * @param {Download} `params` - The parameters for the download request.
     * @param {Download['fileId']} `params.fileId` - The unique identifier of the file to download.
     * @param {Download['sessionId']} `params.sessionId` - The session identifier required for authentication.
     * @returns {Promise<DownloadResponse>} `DownloadResponse` - A promise that resolves with the download response.
     * @throws {Error} `Error` - Throws an error if API initialization is required.
     */
    async download({ fileId, sessionId }) {
        // Log the start of the download process
        this._logger.debug(types_1.EVENT_TYPE.DOWNLOAD, "Running download file.");
        // Check if API initialization is required
        if (!this._checkInitialization()) {
            throw "API Initialization Required: Please initialize the API first.";
        }
        // Construct the download parameters
        const downloadParams = {
            type: types_1.EVENT_TYPE.DOWNLOAD,
            sessionid: sessionId,
            fileid: fileId,
        };
        // Log the start of the download request
        this._logger.debug(types_1.EVENT_TYPE.DOWNLOAD, "Sending download request.");
        // Make the download request and await the response
        const response = await this._makeRequest({
            params: { ...downloadParams },
        });
        // Log the completion of the download request
        this._logger.debug(types_1.EVENT_TYPE.DOWNLOAD, "Download request complete, Returning response.");
        // Return the download response
        return { ...response };
    }
}
exports.default = Api;
