// @keyauthjs/client/dist/types/index.js
"use strict";
// Interfaces for various data structures
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERROR_CODE = exports.EVENT_TYPE = exports.KeyauthEventEmitter = void 0;
const events_1 = __importDefault(require("events"));
class KeyauthEventEmitter extends events_1.default {
    constructor() {
        super();
    }
}
exports.KeyauthEventEmitter = KeyauthEventEmitter;
// Enum defining various event types
var EVENT_TYPE;
(function (EVENT_TYPE) {
    EVENT_TYPE["INIT"] = "init";
    EVENT_TYPE["LOG_IN"] = "login";
    EVENT_TYPE["LOG_OUT"] = "logout";
    EVENT_TYPE["REGISTER"] = "register";
    EVENT_TYPE["LICENSE"] = "license";
    EVENT_TYPE["FETCH_STATS"] = "fetchStats";
    EVENT_TYPE["BAN"] = "ban";
    EVENT_TYPE["CHANGE_USERNAME"] = "changeUsername";
    EVENT_TYPE["CHECK_BLACKLIST"] = "checkblacklist";
    EVENT_TYPE["CHECK"] = "check";
    EVENT_TYPE["DOWNLOAD"] = "file";
    EVENT_TYPE["FETCH_ONLINE"] = "fetchOnline";
    EVENT_TYPE["FORGOT_PASSWORD"] = "forgot";
    EVENT_TYPE["CHAT_GET"] = "chatget";
    EVENT_TYPE["GET_VAR"] = "getvar";
    EVENT_TYPE["LOG"] = "log";
    EVENT_TYPE["CHAT_SEND"] = "chatsend";
    EVENT_TYPE["SET_VAR"] = "setvar";
    EVENT_TYPE["UPGRADE"] = "upgrade";
    EVENT_TYPE["WEBHOOK"] = "webhook";
    EVENT_TYPE["VAR"] = "var";
    // Added custom types
    EVENT_TYPE["REQUEST"] = "request";
    EVENT_TYPE["RESPONSE"] = "response";
    EVENT_TYPE["METADATA"] = "metadata";
    EVENT_TYPE["ERROR"] = "error";
    EVENT_TYPE["SESSION"] = "session";
    EVENT_TYPE["INSTANCE"] = "instance";
    EVENT_TYPE["RATE_LIMIT"] = "ratelimit";
})(EVENT_TYPE || (exports.EVENT_TYPE = EVENT_TYPE = {}));
var ERROR_CODE;
(function (ERROR_CODE) {
    ERROR_CODE["K_SK"] = "seesionKilled";
    ERROR_CODE["K_NSID"] = "noSessionID";
    ERROR_CODE["K_NI"] = "notInitialized";
    ERROR_CODE["K_NLI"] = "notLoggedIn";
    ERROR_CODE["K_UVT"] = "unsupportedVarType";
    ERROR_CODE["K_U"] = "unknown";
    ERROR_CODE["K_NCC"] = "noChatChannel";
    ERROR_CODE["K_ICA"] = "invalidClientApi";
})(ERROR_CODE || (exports.ERROR_CODE = ERROR_CODE = {}));
