// @keyauthjs/client/dist/utils/rateLimiter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RateLimiter = void 0;
/**
 * RateLimiter is a class that provides rate limiting functionality based on the Token Bucket algorithm.
 * It allows you to control the rate at which requests are allowed to be made.
 */
class RateLimiter {
    /**
     * Creates a RateLimiter instance.
     *
     * @param {Object} options - Configuration options for the RateLimiter.
     * @param {number} options.maxTokens - The maximum number of tokens in the bucket.
     * @param {number} options.refillRate - The rate at which tokens are refilled (tokens per millisecond).
     * @throws {Error} - Throws an error if `refillRate` or `maxTokens` is not greater than zero.
     */
    constructor({ maxTokens, refillRate, }) {
        if (refillRate <= 0 || maxTokens <= 0) {
            throw new Error("Refill rate and max tokens must be greater than zero.");
        }
        this.tokens = maxTokens;
        this.lastRefillTime = Date.now();
        this.refillRate = refillRate;
        this.maxTokens = maxTokens;
    }
    /**
     * Refills the token bucket based on the elapsed time.
     */
    refill() {
        const now = Date.now();
        const elapsedTime = now - this.lastRefillTime;
        const tokensToAdd = Math.floor(elapsedTime / this.refillRate);
        this.tokens = Math.min(this.tokens + tokensToAdd, this.maxTokens);
        this.lastRefillTime = now;
    }
    /**
     * Checks if the rate limit has been hit.
     *
     * @returns {boolean} - True if the rate limit has been hit, otherwise false.
     */
    hasHitRateLimit() {
        this.refill();
        if (this.tokens > 0) {
            this.tokens--;
            return false; // Not hit rate limit
        }
        return true; // Hit rate limit
    }
    /**
     * Calculates the time (in milliseconds) until a request can be made.
     *
     * @returns {number} - The time until a request can be made.
     */
    getTimeUntilCanMakeRequest() {
        this.refill();
        if (this.tokens > 0) {
            return 0; // Token is available now
        }
        const now = Date.now();
        const elapsedTime = now - this.lastRefillTime;
        const timeUntilAvailable = Math.max(0, this.refillRate - elapsedTime);
        return timeUntilAvailable;
    }
    /**
     * Asynchronously waits until a request can be made based on the rate limit.
     */
    async sleep(ms) {
        return new Promise((resolve) => setTimeout(resolve, ms));
    }
    /**
     * Asynchronously waits until a request can be made and resets the token count to the maximum.
     */
    async waitUntilCanMakeRequest() {
        const timeToWait = this.getTimeUntilCanMakeRequest();
        if (timeToWait > 0) {
            await this.sleep(timeToWait);
        }
        // After waiting, reset the token count to the maximum
        this.tokens = this.maxTokens;
    }
    /**
     * Returns a human-readable string indicating the time until a request can be made.
     *
     * @returns {string} - A human-readable time string.
     */
    getTimeUntilCanMakeRequestString() {
        const timeUntilAvailable = this.getTimeUntilCanMakeRequest();
        if (timeUntilAvailable === 0) {
            return "Now"; // Token is available now
        }
        if (timeUntilAvailable < 1000) {
            return `${timeUntilAvailable} milliseconds`;
        }
        const seconds = Math.floor(timeUntilAvailable / 1000);
        if (seconds < 60) {
            return `${seconds} second${seconds !== 1 ? "s" : ""}`;
        }
        const minutes = Math.floor(seconds / 60);
        if (minutes < 60) {
            return `${minutes} minute${minutes !== 1 ? "s" : ""}`;
        }
        const hours = Math.floor(minutes / 60);
        return `${hours} hour${hours !== 1 ? "s" : ""}`;
    }
}
exports.RateLimiter = RateLimiter;
